src/persist/src/indexed/columnar/parquet.rs

```diff
@@ -124,25 +137,69 @@ fn encode_parquet_kvtd<W: Write>(
 
 fn decode_parquet_file_kvtd<R: Read + Seek>(
     r: &mut R,
+    format_metadata: Option<ProtoFormatMetadata>,
     metrics: &ColumnarMetrics,
-) -> Result<Vec<ColumnarRecords>, Error> {
+) -> Result<BlobTraceUpdates, Error> {
     let metadata = read_metadata(r)?;
     let schema = infer_schema(&metadata)?;
-    let reader = FileReader::new(r, metadata.row_groups, schema, None, None, None);
-
-    let file_schema = reader.schema().fields.as_slice();
-    // We're not trying to accept any sort of user created data, so be strict.
-    if file_schema != SCHEMA_ARROW_KVTD.fields {
-        return Err(format!(
-            "expected arrow schema {:?} got: {:?}",
-            SCHEMA_ARROW_KVTD.fields, file_schema
-        )
-        .into());
-    }
 
-    let mut ret = Vec::new();
-    for batch in reader {
-        ret.push(decode_arrow_batch_kvtd(&batch?, metrics)?);
+    let mut reader = FileReader::new(r, metadata.row_groups, schema, None, None, None);
+    let file_schema = reader.schema().fields.as_slice().to_vec();
+
+    match format_metadata {
+        None => {
+            // We're not trying to accept any sort of user created data, so be strict.
+            if file_schema != SCHEMA_ARROW_KVTD.fields {
+                return Err(format!(
+                    "expected arrow schema {:?} got: {:?}",
+                    SCHEMA_ARROW_KVTD.fields, file_schema
+                )
+                .into());
+            }
+
+            let mut ret = Vec::new();
+            for batch in reader {
+                let batch = batch?;
+                let records = decode_arrow_batch_kvtd(batch.columns(), metrics)?;
+                ret.push(records);
+            }
+            Ok(BlobTraceUpdates::Row(ret))
+        }
+        Some(ProtoFormatMetadata::StructuredMigration(1)) => {
+            if file_schema.len() > 6 {
+                return Err(
+                    format!("expected at most 6 columns, got {}", file_schema.len()).into(),
+                );
+            }
+
+            let batch = reader
+                .next()
+                .ok_or_else(|| Error::String("found empty batch".to_string()))??;
+            // We enforce an invariant that we have a single RowGroup.
+            if reader.next().is_some() {
+                return Err(Error::String("found more than one RowGroup".to_string()));
+            }
+
+            let columns = batch.columns();
+
+            // The first 4 columns are our primary (K, V, T, D) and optionally
+            // we also have K_S and/or V_S if we wrote structured data.
+            //
+            // TODO(parkmycar): Add some more validation here.
```

Comment:
If you've got something in mind, maybe include more detail in the comment

Reply:
Ping

Reply:
Whoops! I just went ahead and added the validation that I could think of